rudder-core/src/main/scala/com/normation/rudder/services/policies/SystemVariableService.scala

```diff
@@ -375,10 +389,33 @@ class SystemVariableServiceImpl(
         addWithSubChildren(children)
       }
 
-      val varSubNodesName    = systemVariableSpecService.get("SUB_NODES_NAME"   ).toVariable(allChildren.map(_.hostname))
-      val varSubNodesId      = systemVariableSpecService.get("SUB_NODES_ID"     ).toVariable(allChildren.map(_.id.value))
-      val varSubNodesServer  = systemVariableSpecService.get("SUB_NODES_SERVER" ).toVariable(allChildren.map(_.policyServerId.value))
-      val varSubNodesKeyhash = systemVariableSpecService.get("SUB_NODES_KEYHASH").toVariable(allChildren.map(n => s"sha256:${n.sha256KeyHash}"))
+      // Each node may have several agent type, so we need to "unfold" agents per children
+      val subNodesList = allChildren.map(node => (node.agentsName.map(agent => agent -> node))).flatten
+
+      val varSubNodesName    = systemVariableSpecService.get("SUB_NODES_NAME"   ).toVariable(subNodesList.map(_._2.hostname))
+      val varSubNodesId      = systemVariableSpecService.get("SUB_NODES_ID"     ).toVariable(subNodesList.map(_._2.id.value))
+      val varSubNodesServer  = systemVariableSpecService.get("SUB_NODES_SERVER" ).toVariable(subNodesList.map(_._2.policyServerId.value))
+      val varSubNodesKeyhash = systemVariableSpecService.get("SUB_NODES_KEYHASH").toVariable(subNodesList.map(n => s"sha256:${n._2.sha256KeyHash}"))
+
+
+      // Construct the system variables for nodes with certificates
+      val nodesWithCertificate = nodesAgentWithCertificate.flatMap {case (agent, node) =>
+        agent.securityToken match {
+          // A certificat, we return it
+          case cert:Certificate => Some(node -> cert)
```

Comment:
you should return here (cert, cert.cert) because parsing cert is expensive, so better doing only one time